### PR TITLE
685 layer library

### PR DESCRIPTION
**Pull request by redhog**
Connects https://github.com/SkyTruth/pelagos-server/issues/685 and https://github.com/SkyTruth/pelagos-server/issues/908

### CODE REVIEW

File: js/app/UrlValues.js [hunk 1]
```diff
@@ -136,5 +136,23 @@ define(["app/Class", "app/LangExtensions"], function(Class) {
...
+    /* Resolves url relative to a base url */
+    if (url.indexOf("://") != -1) {
+      return url;
+    }
+
+    if (url.indexOf("/") == 0) {
+      var parser = document.createElement('a');
+      parser.href = base;
+
+      base = parser.protocol + "//" + parser.host;
```
Comment (andres-arana) [inline js/app/UrlValues.js:149]: Does the protocol include the `:` character?

File: js/app/Visualization/UI/AnimationLibrary.js [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+
+      $(self.dialog.containerNode).find('.search-loading').show();
+      self.dataManager.listAvailableSources(function (err, sources) {
+        self.sources = sources;
+        var results = [];
+        if (!query) {
+          results = Object.keys(sources);
+        } else {
+          for (var name in sources) {
+            if (name.indexOf(query) != -1) {
```
Comment (andres-arana) [inline js/app/Visualization/UI/AnimationLibrary.js:92]: I think that we should at least normalize (lowercase + maybe remove symbols) before comparing to get better matches.

File: js/app/Visualization/UI/AnimationLibrary.js [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+              results.push(name);
+            } else {
+              if (sources[name].tags) {
+                for (var i = 0; i < sources[name].tags.length; i++) {
+                  if (sources[name].tags[i].indexOf(query)) {
```
Comment (andres-arana) [inline js/app/Visualization/UI/AnimationLibrary.js:97]: I think we should normalize (lowercase + maybe remove symbols) before comparing here as well.

File: js/app/Visualization/UI/AnimationLibrary.js [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+              }
+            }
+          }
+        }
+        results.sort();
+
+        $(self.sourcesList.containerNode).html('');
+        $(self.animationsList.containerNode).html('');
+
+        results.map(function (key) {
```
Comment (andres-arana) [inline js/app/Visualization/UI/AnimationLibrary.js:111]: `forEach` is more appropriate than `map` here, as you are doing stateful transformations of the container node.

File: js/app/UrlValues.js (hunk 1 above)
Comment (redhog) [inline js/app/UrlValues.js:149]:
```
var parser = document.createElement('a');
parser.href = "http://example.com"
parser.protocol
> "http:"
```

So, surprisingly yes

File: js/app/Visualization/UI/AnimationLibrary.js (hunk 4 above)
Comment (redhog) [inline js/app/Visualization/UI/AnimationLibrary.js:111]: forEach doesn't exist everywhere?

Comment (andres-arana) [inline js/app/Visualization/UI/AnimationLibrary.js:111]: It seems from CanIUse that it's supported everywhere maps is supported, as they belong to the same spec. It's a very minor thing though, just something to take into account when writing transformations and sequence processing code. I'll merge it as it is.

**Pull request merged**